### ISSUE TITLE
Compressed MHA writing is slow

### ISSUE THREAD

**Issue by jmsmkn** [DIAGNijmegen/rse-panimg#9]
See if SimpleITK have added support for threading yet

https://github.com/comic/grand-challenge.org/issues/1240

**Comment by nlessmann** [comment]
Loosely related question: are there any plans to support reading image (in particular DICOM) and returning image objects instead of writing the images as mha files to the output folder?

**Comment by jmsmkn** [comment]
No plans, but if that is something that would be used then we can work the use case up into a pitch (https://github.com/DIAGNijmegen/rse-roadmap). We would need some requirements as to how it would be used elsewhere.

**Comment by nlessmann** [comment]
The use case for me would be mainly the integration of the DICOM image builder into the [Tiger](https://github.com/DIAGNijmegen/msk-tiger) library. The `read_dicom` function there is frequently used by me and several other people for reading DICOM files, and we are also working on a [data preparation pipeline](https://github.com/DIAGNijmegen/diag-radiology-preprocessing) that is supposed to read DICOM and do several normalization and segmentation steps before storing the images as mha files. At the moment, how DICOM files are read depends on where the code is executed: inside the standard docker base image on SOL the DIAG dicom loader extension is available and is used, but when this is not available the fallback is SimpleITK. Ideally, I would want to get rid of this dual-backend setup and use the panimg library everywhere. But it would be quite inefficient if panimg would read the DICOM files, write a mha file to disk and the code wrapped around it would immediately read that same mha file again to have the image in memory.

On the other hand, there might also be scenarios where this could lead to problems, e.g., sometimes all series from a DICOM study are dumped into one folder, so reading from that folder would mean loading multiple series into memory. Writing them to disk and reading them later one by one might then be better to avoid memory issues. Although I doubt that this would often be an issue.

In any case, it would be straightforward to work around this with a temporary output directory, so it's really only an efficiency thing - so if it would be very complicated and would require a lot of refactoring, it might not be worth the effort.

In terms of requirements it would be sufficient to return SimpleITK image objects instead of paths to the files.

**Comment by jmsmkn** [comment]
I have made a prototype on the `builder_generators` branch, which changes the builders to accept a set of files, and then yield images, finally raising an exception if any of the files could not be imported.

It would be used like this **(Note: API not final, this might not work)**:

```python
files = {f for f in Path("/foo/").glob("*.dcm) if f.is_file()}

try:
    for result in image_builder_dicom(files=files):
        sitk_image = result.image
        process(sitk_image)  # etc. you can also look at result.name for the name of the file, and result.consumed_files to see what files were used for this image
except BuilderError as e:
    # e.errors will contain a Dict with a list of all the errors for the files that could not be consumed by the builder
```

I don't really know how else to get the file consumption errors out of a generator.

**Comment by nlessmann** [comment]
Great! 👍

**Comment by jmsmkn** [comment]
I'm not very keen on adding this, the vast majority of upload sessions on grand challenge only produce one file so there would be no gains there. The user can split their images up into multiple sessions which we then import in parallel. Consumers of this library can work with the generated images implemented in #19 if they want to implement threaded saving.
